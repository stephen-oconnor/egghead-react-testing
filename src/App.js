import React, { Component } from "react";
import logo from "./logo.svg";
import "./App.css";

const Test = () => <div>Testing</div>;
const Title = ({ text }) => <div>{text}</div>;

class App extends Component {
	constructor(props) {
		super(props);

		this.state = {
			on: false,
			input: "",
			mainColor: "blue",
			lifeCycle: "",
		};
	}

	componentDidMount() {
		this.setState({ lifeCycle: "componentDidMount" });
	}

	handleStrings(str) {
		if (str === "Hello World") {
			return true;
		} else {
			return false;
		}
	}

	render() {
		return (
			<div className="App">
				<header className="App-header">
					<img src={logo} className="App-logo" alt="logo" />
					<h1>React</h1>
					<Title text="Hello" />
					<p>
						Edit <code>src/App.js</code> and save to reload.
					</p>
					<a
						className="App-link"
						href="https://reactjs.org"
						target="_blank"
						rel="noopener noreferrer"
					>
						Learn React
					</a>
					<ul className="test">
						<li>One</li>
						<li>Two</li>
						<li>Three</li>
					</ul>
					<p className={"button-state"}>{this.state.on ? "Yes!" : "No!"}</p>
					<button onClick={() => this.setState({ on: true })}>Click</button>
					<h2>{this.state.input}</h2>
					<input
						type="text"
						onChange={(event) =>
							this.setState({ input: event.currentTarget.value })
						}
					/>
					<h3 className={this.state.mainColor}>Yo</h3>
					<p className={"lifeCycle"}>{this.state.lifeCycle}</p>
				</header>
				<Test />
			</div>
		);
	}
}

export class Link extends Component {
	render() {
		return this.props.hide ? null : <a href={this.props.address}>Click</a>;
	}
}

export default App;
