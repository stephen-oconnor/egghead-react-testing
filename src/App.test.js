import React from "react";
import App, { Link } from "./App";
import { configure, shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import toJson from "enzyme-to-json";

// Enzyme adapter needs to be setup for React 16.
configure({ adapter: new Adapter() });

// Wrap tests in a describe block. <App />
describe("<App /> shallow rendering", () => {
	// shallow is used to test individual components, doesn't care about child components.
	const wrapper = shallow(<App />);

	// Individual test(s).
	it("should render App", () => {
		console.log(wrapper.debug());
	});

	// Find a DOM element.
	it("should contain 1 a element", () => {
		expect(wrapper.find("a").length).toBe(1);
	});

	// Check if a DOM element exists.
	it("a element should have a class of App-link", () => {
		expect(wrapper.find(".App-link").exists()).toBe(true);
	});

	// Scope test to parent --> child hierarchy.
	it("should contain 3 list items", () => {
		expect(wrapper.find("ul").children().length).toBe(3);
	});

	// Check for a class name.
	it("should have a class of test", () => {
		expect(wrapper.find("ul").hasClass("test")).toBe(true);
	});

	// Check for text.
	it("should have an h1 with 'React' text", () => {
		expect(wrapper.find("h1").text()).toBe("React");
	});

	// Check for a prop.
	it("should have a text prop with 'Hello' text", () => {
		expect(wrapper.find('[text="Hello"]').length).toBe(1);
	});

	// Check for an attribute.
	it("should have an alt attribute of 'logo'", () => {
		expect(wrapper.find({ alt: "logo" }).length).toBe(1);
	});

	// Snapshot. A 'picture' of the components structure. Can be updated (u) when necessary.
	it("matches the snapshot", () => {
		const tree = shallow(<App />);
		expect(toJson(tree)).toMatchSnapshot();
	});

	// Simulate button click.
	it("on button click changes p text", () => {
		const button = wrapper.find("button");
		expect(wrapper.find(".button-state").text()).toBe("No!");
		// Get the onClick prop and call the function associated.
		button.simulate("click");
		expect(wrapper.find(".button-state").text()).toBe("Yes!");
	});

	// Simulate input change.
	it("on input change, title changes text", () => {
		const input = wrapper.find("input");
		expect(wrapper.find("h2").text()).toBe("");
		// Get the onChange props and call the function associated. Pass in optional event Object.
		input.simulate("change", { currentTarget: { value: "Stephen" } });
		expect(wrapper.find("h2").text()).toBe("Stephen");
	});

	// Check for state updates.
	it("updates className with new state", () => {
		expect(wrapper.find(".blue").length).toBe(1);
		expect(wrapper.find(".red").length).toBe(0);
		wrapper.setState({ mainColor: "red" });
		expect(wrapper.find(".red").length).toBe(1);
		expect(wrapper.find(".blue").length).toBe(0);
	});

	// Check lifecycle methods.
	it("calls componentDidMount, updates p tag", () => {
		expect(wrapper.find(".lifeCycle").text()).toBe("componentDidMount");
	});

	// Check function call return.
	it("handleStrings function returns correctly", () => {
		const trueReturn = wrapper.instance().handleStrings("Hello World");
		const falseReturn = wrapper.instance().handleStrings("");
		expect(trueReturn).toBe(true);
		expect(falseReturn).toBe(false);
	});
});

// Wrap tests in a describe block. <App />
describe("<App /> mount rendering", () => {
	// mount will render the full DOM.
	const wrapper = mount(<App />);

	// Check for text.
	it("should have an h1 with 'React' text", () => {
		expect(wrapper.find("h1").text()).toBe("React");
		wrapper.unmount();
	});

	// Snapshot. A 'picture' of the components structure. Can be updated (u) when necessary.
	it("matches the snapshot", () => {
		const tree = shallow(<App />);
		expect(toJson(tree)).toMatchSnapshot();
		tree.unmount();
	});
});

// Wrap tests in a describe block. <Link />
describe("<Link />", () => {
	// Check instance value of prop.
	it("link component accepts address prop", () => {
		const wrapper = shallow(<Link address="www.google.com" />);
		expect(wrapper.instance().props.address).toBe("www.google.com");
	});

	// Check a tag is using correct href value.
	it("a tag node renders href correctly", () => {
		const wrapper = shallow(<Link address="www.google.com" />);
		expect(wrapper.props().href).toBe("www.google.com");
		// Check when props are updated. Causes a re-render.
		wrapper.setProps({ hide: true });
		expect(wrapper.get(0)).toBeNull();
	});
});
