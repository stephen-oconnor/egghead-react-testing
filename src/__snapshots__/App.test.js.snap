// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> mount rendering matches the snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1>
      React
    </h1>
    <Title
      text="Hello"
    />
    <p>
      Edit 
      <code>
        src/App.js
      </code>
       and save to reload.
    </p>
    <a
      className="App-link"
      href="https://reactjs.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Learn React
    </a>
    <ul
      className="test"
    >
      <li>
        One
      </li>
      <li>
        Two
      </li>
      <li>
        Three
      </li>
    </ul>
    <p
      className="button-state"
    >
      No!
    </p>
    <button
      onClick={[Function]}
    >
      Click
    </button>
    <h2 />
    <input
      onChange={[Function]}
      type="text"
    />
    <h3
      className="blue"
    >
      Yo
    </h3>
    <p
      className="lifeCycle"
    >
      componentDidMount
    </p>
  </header>
  <Test />
</div>
`;

exports[`<App /> shallow rendering matches the snapshot 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1>
      React
    </h1>
    <Title
      text="Hello"
    />
    <p>
      Edit 
      <code>
        src/App.js
      </code>
       and save to reload.
    </p>
    <a
      className="App-link"
      href="https://reactjs.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Learn React
    </a>
    <ul
      className="test"
    >
      <li>
        One
      </li>
      <li>
        Two
      </li>
      <li>
        Three
      </li>
    </ul>
    <p
      className="button-state"
    >
      No!
    </p>
    <button
      onClick={[Function]}
    >
      Click
    </button>
    <h2 />
    <input
      onChange={[Function]}
      type="text"
    />
    <h3
      className="blue"
    >
      Yo
    </h3>
    <p
      className="lifeCycle"
    >
      componentDidMount
    </p>
  </header>
  <Test />
</div>
`;
